### PR TITLE
[ADD] ir_actions.py: logger log in server action in context

### PR DESCRIPTION
**Pull request by lse-odoo**
This PR is mainly meant for the support.

Note that the `stack_info` is purposefully accessible to add some stack trace in the logs.

Example of support tickets where it can be useful:
A certain field of a particular model change it's value with no particular pattern and way to reproduce.
With this commit, we can now create an automated actions on the model update trigger to dump the current stack that will lead us on the action that did trigger it.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/addons/base/models/ir_actions.py [hunk 1]
```diff
@@ -550,6 +550,9 @@ def log(message, level="info"):
                     VALUES (NOW() at time zone 'UTC', %s, %s, %s, %s, %s, %s, %s, %s, %s)
                 """, (self.env.uid, 'server', self._cr.dbname, __name__, level, message, "action", action.id, action.name))
 
+        def logger_log(message="", stack_info=False, level=20):
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_actions.py:553]:
```suggestion
        def logger_log(message="", stack_info=False, level=logging.INFO):
```

**Review by rco-odoo — APPROVED** [review]
Looks fine to me.

File: odoo/addons/base/models/ir_actions.py (hunk 1 above)
Comment (lse-odoo) [reply odoo/addons/base/models/ir_actions.py]: indeed, thanks :)

**Comment by william-andre** [comment]
I have this, which is similar https://github.com/odoo/odoo/pull/70641
And used here  https://github.com/odoo-dev/enterprise/commit/78f779cac0bf25b8cca2ad057388165aa42e122b#diff-11f77384cf35f932d5252da45f6107e5e5e146eca3b2d9a26acc7d076a5d82b6R70-R74

File: odoo/addons/base/models/ir_actions.py [hunk 2]
```diff
@@ -389,6 +411,7 @@ class IrActionsServer(models.Model):
 #  - time, datetime, dateutil, timezone: useful Python libraries
 #  - float_compare: Odoo function to compare floats based on specific precisions
 #  - log: log(message, level='info'): logging function to record debug information in ir.logging table
+#  - _logger: _logger.info(message): logging class to record debug information in Odoo's log
```
Comment (Xavier-Do) [inline odoo/addons/base/models/ir_actions.py:414]: Are we sure we want to document this one?

Comment (lse-odoo) [reply odoo/addons/base/models/ir_actions.py:414]: > Are we sure we want to document this one?

> "oui a priori si c'est visible dans le code, autant que ce soit documenté" (ODO)

See: https://discord.com/channels/678381219515465750/687339689522364423/877929059009577021

**Review by Xavier-Do — APPROVED** [review]

**Comment by Xavier-Do** [comment]
@robodoo r+

File: odoo/addons/base/models/ir_actions.py [hunk 3]
```diff
@@ -16,8 +16,30 @@
...
+    def info(message, *args, stack_info=False, exc_info=False):
+        _server_action_logger.info(message, *args, stack_info=stack_info, exc_info=exc_info)
+
+    @staticmethod
+    def warning(message, *args, stack_info=False, exc_info=False):
+        _server_action_logger.warning(message, *args, stack_info=stack_info, exc_info=exc_info)
+
+    @staticmethod
+    def error(message, *args, stack_info=False, exc_info=False):
+        _server_action_logger.error(message, args, stack_info=stack_info, exc_info=exc_info)
```
Comment (odony) [inline odoo/addons/base/models/ir_actions.py:37]:
```suggestion
        _server_action_logger.error(message, *args, stack_info=stack_info, exc_info=exc_info)
```

File: odoo/addons/base/models/ir_actions.py [hunk 4]
```diff
@@ -16,8 +16,30 @@
...
+
+    @staticmethod
+    def exception(message, *args, stack_info=False, exc_info=True):
+        _server_action_logger.exception(message, args, stack_info=stack_info, exc_info=exc_info)
```
Comment (odony) [inline odoo/addons/base/models/ir_actions.py:41]:
```suggestion
        _server_action_logger.exception(message, *args, stack_info=stack_info, exc_info=exc_info)
```

**Comment by odony** [comment]
I added a small commit for spellchecking/sync'ing the inline doc, as it has been getting uglier and out of sync as time goes. If you prefer me to do that in a different PR, that's fine, just drop my commit and let me know ;)

**Comment by lse-odoo** [comment]
@odony i just noticed that this PR never ended up being merged :/ do I need to do something in particular on my side (maybe add a test like you proposed ?) & solve the conflict ?

File: odoo/addons/base/models/ir_actions.py [hunk 5]
```diff
@@ -17,8 +17,30 @@
 from pytz import timezone
 
 _logger = logging.getLogger(__name__)
+_server_action_logger = _logger.getChild("server_action_safe_eval")
 
 
+class LoggerProxy:
```
Comment (Julien00859) [inline odoo/addons/base/models/ir_actions.py:23]: Can you add a small docstring explaining that we don't want to expose too much of logging.logger as it is used in safe_eval?

**Review by Julien00859 — CHANGES_REQUESTED** [review]
The two commits shouldn't had be squashed together, it makes your actual change hard to isolate from the typos

**Comment by lse-odoo** [comment]
@Julien00859 
We did have a lot of trouble to revive ODO's commit, but we (with the help of MAO) were able to restore some part of it (one part ended up in my commit for some reason).

Let me know what you think of it now :)

**Review by Julien00859 — APPROVED** [review]
@robodoo r+

**Comment by Julien00859** [comment]
> @Julien00859 We did have a lot of trouble to revive ODO's commit, but we (with the help of MAO) were able to restore some part of it (one part ended up in my commit for some reason).
> 
> Let me know what you think of it now :)

The way I would had done it:

1) Cancel the current commit but keep the changes in the working directory `git reset HEAD~1`
2) Verify that you have a "dirty" working directory `git status`
3) Add only the changes of Olivier `git add -p` (-p prompt you for each diff if you want to add it or leave it, take some time)
4) Commit those changes as Olivier `git commit --author "Olivier Dony <odo@odoo.com>"
5) Add the remaining changes and commit them `git commit -a`

The important parts are (1) `git reset HEAD~1`, reset is used to retarget the current branch on a specific commit. With `--hard` it cleans the working directory too: you loose the changes. With `--soft` (default) it doesn't touch the working directory: you keep the changes (you just loose the commits). (2) `git add -p` which allow you to only commit part of the diff.

**Comment by Julien00859** [comment]
@robodoo rebase-ff

**Pull request closed without merging**